spacy/cli/debug_data.py

```diff
@@ -670,6 +671,40 @@ def debug_data(
                 f"Found {gold_train_data['n_cycles']} projectivized train sentence(s) with cycles"
             )
 
+    if "trainable_lemmatizer" in factory_names:
+        msg.divider("Trainable Lemmatizer")
+        trees_train: Set[str] = gold_train_data["lemmatizer_trees"]
+        trees_dev: Set[str] = gold_dev_data["lemmatizer_trees"]
+
+        # This is necessary context when someone is attempting to interpret whether the
+        # number of trees exclusively in the dev set is meaningful.
+        msg.info(f"{len(trees_train)} lemmatizer trees generated from training data.")
+        msg.info(f"{len(trees_dev)} lemmatizer trees generated from dev data.")
+
+        dev_not_train = trees_dev - trees_train
+        if len(dev_not_train) != 0:
+            msg.warn(
+                f"{len(dev_not_train)} lemmatizer trees were found exclusively in the dev data."
```

Comment:
This is pretty much always going to be the case, so I'm not sure this warning is going to be helpful. Is there a threshold of cases that might make more sense, like 5% of edit trees or something?

Reply:
Maybe @danieldk has an opinion on a threshold. Based on my understanding, a threshold without any context is difficult to make sense of. Here's an example he gave from our discussion:

> As an example, suppose that 100 of the trees in the dev set are not in the train set, it doesn’t tell much yet. It could be that those 100 trees are unique because they are all derived from misspelled words and consequently only 100 words could not be lemmatized. Or it could be that those 100 trees are fairly frequent and 1000 words could not be lemmatized.

Maybe the thing to do here is to switch this to an `info` rather than `warn`, I do think it's information a user would want but you're right in that it's almost always going to be triggered.

Reply:
I think at the very least the numbers need more context, because a count alone isn't going to mean much, while something like a percentage (even if it's not a perfect measure) might put it more in perspective. Or something like the percentage of words in the dev set that are affected by missing trees?

Reply:
I agree that it makes most sense to make this an info message. Hard to say what would be a good threshold without knowing more about the underlying properties of the data.

Reply:
I think some form of relative numbers would be good here (percentage of trees if that's all we have, or better percentage of dev tokens).

Reply:
Here's what this looks like as % of trees:

```
ℹ 105 lemmatizer trees (18.6% of dev trees) were found exclusively in
the dev data.
```

In implementing a percentage of words measure, I'm encountering an discrepancy calculating the unique trees and I want to confirm the correct way. (The number of trees above I believe is incorrect)

Right now I'm calculating the string representation of each tree and adding that to a `set` as well as storing the tree strings into a dictionary with the `token.text`. The idea is that if we want to know what % of dev tokens don't have trees, we can map back to `data["words"]`, which is a Counter of words, and sum those values for the keys that are in `dev_not_train`

It looks something like this -

```python
def debug_data(
        ...
        tree_id = trees.add(token.text, token.lemma_)
        tree_str = trees.tree_to_str(tree_id)
        data["lemmatizer_trees_dict"][token.text] = tree_str
        data["lemmatizer_trees_set"].add(tree_str)
```

However, when I do this, I get different values for `len(set(data["lemmatizer_trees_dict"].values()))` and `len(data["lemmatizer_trees_set"])`. Specifically on the demo data I mention in the parent comment:

```
ℹ 2298 lemmatizer trees generated from training data. (dict)
ℹ 528 lemmatizer trees generated from dev data. (dict)
ℹ 2709 lemmatizer trees generated from training data. (set)
ℹ 565 lemmatizer trees generated from dev data. (set)
```

So the `set` always ends up with more. I'm thinking this is because the trees & subtress are recursively updating over time so the string representations change, so at `t1` the tree for `word` is going to be different than at `t2`, but in the case of using `set.add` the tree at `t1` and `t2` are incorrectly added when all we really care about is the value at `t_final` (`t2` in this dummy example). 

Is the `dict` approach right @danieldk?

Reply:
The dictionary approach is incorrect because the same surface form may occur with different parts of speech and thus different lemmas. E.g.:

_Hij **dicht** de de scheur._ (He fixes the crack) -> POS: verb, lemma: dichten
_De deur is **dicht**._ (The door is closed) -> POS: adjective, lemma: dicht

If you'd key the dictionary by surface form and part-of-speech, the count should be much closer to the set count. But I am not sure if that is ok for `debug data`, since the corpus may not actually have POS annotations.

Reply:
I think we should just take the `set` approach then, as I don't think we want to depend on POS annotations, and we can compare sets between training and dev and display the percent.